geopandas/io/file.py

```diff
@@ -83,17 +94,45 @@ def _read_file(filename, bbox=None, mask=None, rows=None, **kwargs):
     """
     if _is_url(filename):
         req = _urlopen(filename)
-        path_or_bytes = req.read()
-        reader = fiona.BytesCollection
-    elif isinstance(filename, io.TextIOBase):
-        path_or_bytes = filename.read()
-        reader = fiona.open
+        data = req.read()
+        if get_filetype(data) == "zip":
+            reader = functools.partial(fiona.BytesCollection, data)
+        else:
+            memfile = fiona.io.MemoryFile(data)
+            reader = memfile.open
```

Comment:
Is there a benefit of using `MemoryFile` for the non-zip cases? (because I understand that BytesCollection also handles those, so that would only be one code path then)

Reply:
It seems there also is a ZipMemoryFile, which I suppose is meant as replacement of BytesCollection, long term (https://fiona.readthedocs.io/en/latest/manual.html#memoryfile-and-zipmemoryfile)

Reply:
I don't think there is a benefit to using `MemoryFile`, except that I saw a comment from @sgillies that `BytesCollection` is eventually [slated to be deprecated](https://github.com/Toblerity/Fiona/issues/500#issue-273161477) in favor of the former. I agree it would be simpler to just have one code path here.

My original intention here was to actually replace `BytesCollection` entirely with `ZipMemoryFile` and `MemoryFile`. However, when I tried to do that, I found that `BytesCollection` was the only place that handled the [automatic zip-file inference](https://github.com/Toblerity/Fiona/blob/e46abf2781a490616e8bc306f2a0ba6d98976ffd/fiona/collection.py#L533-L543) (and naming of the virtual buffer).

Rather than duplicating more of that logic in what is already a bit of a complicated set of branches, I thought it wouldn't be so bad to continue using `BytesCollection` for now. Happy to revisit that, however.

Reply:
If we keep using BytesCollection for the zip case, I would use it for both cases, since that simplifies the code here somewhat. 

But I agree it would maybe be good to actually use the (Zip)MemoryFiles, since those are the recommended way by fiona.

> when I tried to do that, I found that BytesCollection was the only place that handled the automatic zip-file inference (and naming of the virtual buffer).

So simply replacing (after having detected the data is from a zip file) `BytesCollection(data)` with `ZipMemoryFile(data)` doesn't work? That is however what I would expect to work based on the example in the fiona docs: https://fiona.readthedocs.io/en/latest/manual.html#memoryfile-and-zipmemoryfile 
(if something is not working, that might be useful feedback for fiona)

Reply:
> If we keep using BytesCollection for the zip case, I would use it for both cases, since that simplifies the code here somewhat.


For the time being, yeah, may as well fall back on BytesCollection for everything, and leave updating to `*MemoryFile` as a TODO.


> So simply replacing (after having detected the data is from a zip file) `BytesCollection(data)` with `ZipMemoryFile(data)` doesn't work? That is however what I would expect to work based on the example in the fiona docs: https://fiona.readthedocs.io/en/latest/manual.html#memoryfile-and-zipmemoryfile
> (if something is not working, that might be useful feedback for fiona)


Took a closer look at this, and opened Toblerity/Fiona#927 discussing the apparent limitation of `ZipMemoryFile`

Reply:
Ah, I see. Thanks for looking into it and opening an issue!